io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,549 @@
+      } else {
+         Bool_t todelete = kFALSE;
+         if (withReset) {
+            todelete = (0 != cl->GetResetAfterMerge());
+         } else {
+            todelete = (0 == cl->GetResetAfterMerge());
+         }
+         if (todelete) {
+            key->Delete();
+            dir->GetListOfKeys()->Remove(key);
+            delete key;
+         }
+      }
+   }
+}
+
+Bool_t TMPIFile::ParallelFileMerger::R__NeedInitialMerge(TDirectory *dir)
+{
+   if (dir == 0)
+      return kFALSE;
+   TIter nextkey(dir->GetListOfKeys());
+   TKey *key;
+   while ((key = (TKey *)nextkey())) {
+      TClass *cl = TClass::GetClass(key->GetClassName());
+      if (cl->InheritsFrom(TDirectory::Class())) {
+         TDirectory *subdir = (TDirectory *)dir->GetList()->FindObject(key->GetName());
+         if (!subdir) {
+            subdir = (TDirectory *)key->ReadObj();
+         }
+         if (R__NeedInitialMerge(subdir)) {
+            return kTRUE;
+         }
+      } else {
+         if (0 != cl->GetResetAfterMerge()) {
+            return kTRUE;
+         }
+      }
+   }
+   return kFALSE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Initial merge of the input to copy the resetable object (TTree) into the output
+/// and remove them from the input file.
+///
+/// taken from root/tutorials/net/parallelMergeServer.C
+
+Bool_t TMPIFile::ParallelFileMerger::InitialMerge(TFile *input)
+{
+   // Initial merge of the input to copy the resetable object (TTree) into the
+   // output and remove them from the input file.
+   fMerger.AddFile(input);
+   Bool_t result =
+      fMerger.PartialMerge(TFileMerger::kIncremental | TFileMerger::kResetable | TFileMerger::kKeepCompression);
+   R__DeleteObject(input, kTRUE);
+   return result;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Merge the current inputs into the output file.
+///
+/// taken from root/tutorials/net/parallelMergeServer.C
+
+Bool_t TMPIFile::ParallelFileMerger::Merge()
+{
+   // Remove object that can *not* be incrementally merge and will *not* be reset by the client code.
+   R__DeleteObject(fMerger.GetOutputFile(), kFALSE);
+   for (UInt_t f = 0; f < fClients.size(); ++f) {
+      fMerger.AddFile(fClients[f].GetFile());
+   }
+   Bool_t result = fMerger.PartialMerge(TFileMerger::kAllIncremental | TFileMerger::kKeepCompression);
+
+   // Remove any 'resetable' object (like TTree) from the input file so that they
+   // will not be re-merged.  Keep only the object that always need to be
+   // re-merged (Histograms).
+   for (UInt_t f = 0; f < fClients.size(); ++f) {
+      if (fClients[f].GetFile()) {
+         R__DeleteObject(fClients[f].GetFile(), kTRUE);
+      } else {
+         // We back up the file (probably due to memory constraint)
+         TFile *file = TFile::Open(fClients[f].GetLocalName(), "UPDATE");
+         if (file->IsZombie())
+            exit(1);
```

Comment:
This is too harsh :).   Please find issue an Error message and return false (after doing the necessary cleanups).

Reply:
done!